ui/src/views/Documentation/index.jsx

```diff
@@ -106,63 +182,220 @@ export default class Documentation extends Component {
+    return {
+      a: this.anchorFactory,
+      h1: this.headingFactory('h1'),
+      h2: this.headingFactory('h2'),
+      h3: this.headingFactory('h3'),
+      h4: this.headingFactory('h4'),
+      h5: this.headingFactory('h5'),
+      h6: this.headingFactory('h6'),
+      img: this.imageFactory,
+    };
+  }
+
   async load() {
     try {
-      const { default: Document } = await this.handleImport(
-        this.props.match.params.path
+      const { params } = this.props.match;
+      const pathname = params.path || 'index';
+      const isJsonFile =
+        pathname.startsWith('reference') &&
+        (pathname.endsWith('api') || pathname.endsWith('events'));
+      const { loader, path } = readDocFile(
+        `${pathname}.${isJsonFile ? 'json' : 'md'}`
       );
+      const { default: Page } = await loader;
+
+      absolutePath = path;
+
+      // use `isMDXComponent` once https://github.com/mdx-js/mdx/pull/369 is merged
+      if (!Page.prototype) {
+        const entries = await this.getReferenceEntries(Page.entries);
+
+        return this.setState({
+          Page: null,
+          pageInfo: null,
+          error: null,
+          referenceJson: Object.assign({}, Page, { entries }),
+        });
+      }
+
       const pageInfo = this.getPageInfo();
 
-      this.setState({ Document, pageInfo, error: null });
+      this.setState({ Page, pageInfo, error: null, referenceJson: null });
     } catch (error) {
       this.setState({ error });
     }
   }
 
-  async handleImport(url) {
-    const doc = url ? url.replace(/\/$/, '') : 'index';
+  getDocument(Page) {
+    if (!Page) {
+      return null;
+    }
 
-    return import(/* webpackChunkName: 'Documentation.page' */ `../../docs/${doc}.md`).catch(
-      () =>
-        import(/* webpackChunkName: 'Documentation.page' */ `../../docs/${doc}/index.md`)
+    const page = renderToString(
+      <BrowserRouter>
+        <Page components={this.components()} />
+      </BrowserRouter>
     );
+
+    return page
+      .split(DOCS_SCHEMA_REGEX)
+      .map(html =>
+        !isUrl(html) && html.startsWith('<') ? (
+          html
+        ) : (
+          <SchemaTable key={html} schema={html} />
+        )
+      );
   }
 
+  getReferenceEntries = entries => {
+    const projectName = this.props.match.params.path.split('/')[2];
+
+    return Promise.all(
+      entries.map(entry =>
+        ['input', 'output', 'schema'].reduce(async (acc, prop) => {
+          const accumulator = await acc;
+
+          if (!(prop in entry)) {
+            return acc;
+          }
+
+          const schemaName = entry[prop].replace(/#$/, '');
+          let {
+            default: schema,
+          } = await import(/* webpackChunkName: 'Documentation.Schema' */ `../../../../generated/docs/${projectName}/schemas/${schemaName}`);
+
+          schema = this.sanitizeSchema(schema);
+
+          // `dereference` maintains object reference equality
+          const deref = await RefParser.dereference(
```

Comment:
Rather than flatten references, I'd like to eventually have react-schema-viewer able to follow references (shown by expanding a collapsed panel).  That will mean feeding the entire universe of schemas to react-schema-viewer and then telling it the ID of the schema it should render.  That universe is currently included in `generated/references.json` -- it's not huge (69K gzipped).  So, we might want to look forward to a day when all schemas are in a single data structure in memory. For now, I think this is OK..